### PR TITLE
refactor: use TableManager to implement `CatalogManager` and `SchemaManager`

### PR DESCRIPTION
**Pull request by waynexia**
Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

# Which issue does this PR close?

Part of #245.

# Rationale for this change
 

`CatalogManager` and `SchemaManager`'s interfaces are redundant in some sorts. This is one step towards `Cluster` and `catalog`'s combination. The next step should be to implement table manipulating operations on top of this.

# What changes are included in this PR?

This patch implements `CatalogManager` and `SchemaManager` using `TableManager`, and implement `catalog::manager::Manager` for `ClusterImpl`

The new `catalog::manager::Manager` implementation should replace `volatile`, but I haven't removed the old one to keep this patch small.

# Are there any user-facing changes?

No so far.

# How does this change test

untested.

### CODE REVIEW

File: catalog_impls/src/cluster.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                    table_manager: self.table_manager.clone(),
+                }) as _
+            });
+
+        Ok(schema)
+    }
+
+    #[allow(unused_variables)]
+    async fn create_schema<'a>(&'a self, name: NameRef<'a>) -> catalog::Result<()> {
+        todo!()
```
Comment (ShiKaiWi) [inline catalog_impls/src/cluster.rs:88]: Will `create_schema` be supported in the following patches?

File: cluster/src/table_manager.rs [hunk 2]
```diff
@@ -47,51 +47,138 @@ impl From<&CreateTableCmd> for ShardTableInfo {
...
-            .drop_table(schema_name, table_name)
+            .catalog_infos
+            .keys()
+            .cloned()
+            .collect()
     }
 
-    pub fn update_table_info(&self, shard_table: &HashMap<ShardId, ShardTables>) {
-        self.inner.write().unwrap().update_table_info(shard_table)
+    // todo: should return schema id as well?
```
Comment (ShiKaiWi) [inline cluster/src/table_manager.rs:94]: todo -> TODO

File: cluster/src/table_manager.rs [hunk 3]
```diff
@@ -47,51 +47,138 @@ impl From<&CreateTableCmd> for ShardTableInfo {
...
 #[derive(Debug, Default)]
 struct Inner {
+    // shard infos
     shard_infos: HashMap<ShardId, ShardInfo>,
-    schema_infos: HashMap<SchemaName, SchemaInfo>,
-    // TODO: maybe another mapping (shard -> table) is necessary.
-    tables: BTreeMap<SchemaName, BTreeMap<TableName, ShardTableInfo>>,
+
+    // catalog/schema infos
+    catalog_infos: HashMap<String, HashMap<SchemaName, SchemaInfo>>,
```
Comment (ShiKaiWi) [inline cluster/src/table_manager.rs:157]: Shall we define a type called CatalogName just like SchemaName?

File: cluster/src/table_manager.rs [hunk 4]
```diff
@@ -47,51 +47,138 @@ impl From<&CreateTableCmd> for ShardTableInfo {
...
+
+    // table handles
+    tables_org_by_token: BTreeMap<TableToken, TableRef>,
+    #[allow(dead_code)]
+    shard_to_table_tokens: HashMap<ShardId, HashSet<TableToken>>,
```
Comment (ShiKaiWi) [inline cluster/src/table_manager.rs:162]: How about:
- tables_by_token
- tokens_by_shard

File: cluster/src/table_manager.rs [hunk 5]
```diff
@@ -47,51 +47,138 @@ impl From<&CreateTableCmd> for ShardTableInfo {
...
+}
+
+#[repr(C)]
```
Comment (ShiKaiWi) [inline cluster/src/table_manager.rs:165]: Is this necessary?

File: catalog_impls/src/cluster.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    schema::{
+        self, CloseOptions, CloseTableRequest, CreateOptions, CreateTableRequest, DropOptions,
+        DropTableRequest, NameRef, OpenOptions, OpenTableRequest, Schema, SchemaRef,
+    },
+    Catalog, CatalogRef,
+};
+use cluster::{cluster_impl::ClusterImpl, table_manager::TableManager};
+use table_engine::table::{SchemaId, TableRef};
+
+pub struct ManagerImpl(ClusterImpl);
```
Comment (ShiKaiWi) [inline catalog_impls/src/cluster.rs:16]: Maybe we need some comments to explain this implement.

File: cluster/src/table_manager.rs (hunk 2 above)
Comment (waynexia) [reply cluster/src/table_manager.rs]: Done, this is a stale TODO

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Pull request merged**